src/main/java/org/terasology/moduletestingenvironment/TestEventReceiver.java

```diff
@@ -63,64 +74,79 @@
+public class TestEventReceiver<T extends Event> implements AutoCloseable, EventReceiver<T> {
+    private final EventSystem eventSystem;
+    private final Class<T> eventClass;
+    private final BiConsumer<T, EntityRef> handler;
 
-    private EventSystem eventSystem;
-    private Class<T> eventClass;
-    private List<EntityRef> entityRefs = new ArrayList<>();
-    private List<T> events = new ArrayList<>();
-    private BiConsumer<T, EntityRef> handler = (a, b)-> { };
+    private final List<EntityRef> entityRefs = new ArrayList<>();
+    private final List<T> events = new ArrayList<>();
 
     /**
      * Constructs a new {@code TestEventReceiver} and registers it to listen for events.
-     *
+     * <p>
+     * The following signature of a {@code TestEventReceiver} is equivalent to the event handler below:
      * <pre>
-     * {@literal
-     * TestEventReceiver receiver = new TestEventReceiver<>(context, DropItemEvent.class, (event, entity) -> {
+     * {@code
+     * TestEventReceiver receiver = new TestEventReceiver<>(context, MyEvent.class, (event, entity) -> {
      *   // do something with the event or entity
-     * });
+     * }, MyComponent.class);
+     *
+     * // ... corresponds to
+     *
+     * @ReceiveEvent(components = {MyComponent.class})
+     * public void handler(MyEvent event, EntityRef entity) {
+     *     // do something with the event or entity
+     * }
      * }
      * </pre>
      *
-     * @param context the context object for the test; this should probably be obtained through
-     *                {@link ModuleTestingEnvironment#getHostContext()} and is needed so we can
-     *                obtain an {@link EventSystem} instance to register our event handler.
-     * @param eventClass   the {@link Event} subclass to listen for
-     * @param handler an optional {@link BiConsumer} to fire when events are received.
+     * @param context the context object for the test; this should probably be obtained through {@link
+     *         ModuleTestingHelper#createClient()} and is needed so we can obtain an {@link EventSystem} instance to
+     *         register our event handler.
+     * @param eventClass the {@link Event} subclass to listen for
+     * @param handler an optional {@link BiConsumer} fired when events are received
+     * @param componentTypes list of component types that need to be present on the entity receiving the event
      */
-    public TestEventReceiver(Context context, Class<T> eventClass, BiConsumer<T, EntityRef> handler) {
-        this.handler = handler;
+    public TestEventReceiver(Context context, Class<T> eventClass, BiConsumer<T, EntityRef> handler, Class<?
+            extends Component>... componentTypes) {
         this.eventClass = eventClass;
+        this.handler = handler;
         eventSystem = context.get(EventSystem.class);
-        eventSystem.registerEventReceiver(this, eventClass, EntityInfoComponent.class);
+
+        Class<? extends Component>[] components =
```

Comment:
This line now has a warning about an unsafe conversion:

> Unchecked assignment: 'java.lang.Class[]' to 'java.lang.Class<? extends org.terasology.entitySystem.Component>[]' 

I believe this to be safe in this context. I'd prefer a cleaner solution if anybody can think of one, though 😉

Reply:
This looks perfectly safe since the input types are bounded. I wouldn't run a business off of it but it's fine for this purpose.